### PR TITLE
swarm/shed: add metrics to each shed db

### PR DESCRIPTION
**Pull request by nonsense**
Currently it is kind of difficult to measure Swarm's disk usage, due to the very different setups we use. It'd be nice if we have a consistent view of the disk usage at least from the point of view of the Swarm process.

### CODE REVIEW

File: swarm/shed/db.go [hunk 1]
```diff
@@ -23,26 +23,48 @@
...
+
+	quitLock sync.Mutex      // Mutex protecting the quit channel access
+	quitChan chan chan error // Quit channel to stop the metrics collection before closing the database
 }
 
 // NewDB constructs a new DB and validates the schema
 // if it exists in database on the given path.
-func NewDB(path string) (db *DB, err error) {
+// prefix is used for metrics collection for the given DB.
+func NewDB(path string, prefix string) (db *DB, err error) {
```
[the remark this answers is not in the document]
Comment (nonsense) [reply swarm/shed/db.go]: I think there is no harm calling it for every single LevelDB, as metrics are gathered only once every 10sec. and AFAIK LevelDB also collects them always.

File: swarm/shed/db.go [hunk 2]
```diff
@@ -126,5 +151,182 @@ func (db *DB) WriteBatch(batch *leveldb.Batch) (err error) {
...
 // Close closes LevelDB database.
 func (db *DB) Close() (err error) {
+	close(db.quitChan)
 	return db.ldb.Close()
 }
+
+// Meter configures the database metrics collectors
+func (db *DB) Meter(prefix string) {
+	// Initialize all the metrics collector at the requested prefix
+	db.compTimeMeter = metrics.NewRegisteredMeter(prefix+"compact/time", nil)
```
Comment (janos) [inline swarm/shed/db.go:165]: Shouldn't *Meter fields assignments be under a lock?

File: swarm/shed/db.go [hunk 3]
```diff
@@ -126,5 +151,182 @@ func (db *DB) WriteBatch(batch *leveldb.Batch) (err error) {
...
+				if err != nil {
+					log.Error("Compaction entry parsing failed", "err", err)
+					merr = err
+					continue
+				}
+				compactions[i%2][idx] += value
+			}
+		}
+		// Update all the requested meters
+		if db.compTimeMeter != nil {
```
Comment (janos) [inline swarm/shed/db.go:235]: All *Meter fields checks should be under a lock.

I have a suggestion, not to export Meter method, just to initialize metrics in NewDB, making them "immutable" from "outside" and to remove the lock completely.

**Review by janos — COMMENTED** [review]
Thank Anton, I have anther suggestion. Not from the performance point of view, but to make code a bit more simple by removing the lock.

File: swarm/shed/db.go (hunk 2 above)
Comment (nonsense) [reply swarm/shed/db.go:165]: I don't think so. Why would they be under a lock? `Meter` is called once, when you initialise the DB, why would you initialise the same DB under different threads?

Comment (janos) [reply swarm/shed/db.go:165]: As Meter is exported, it can be called any number of times, this is why I suggested not to export it, in the commend bellow.

Comment (nonsense) [reply swarm/shed/db.go:165]: You're right, `Meter` should not be exported. It should not be called a number of times, but only once.

File: swarm/shed/db.go (hunk 3 above)
Comment (nonsense) [reply swarm/shed/db.go:235]: Yes, the problem is that `Meter` is exported.

**Review by janos — APPROVED** [review]
LGTM. Thanks @nonsense.

**Comment by zelig** [comment]
merged upstream as https://github.com/ethereum/go-ethereum/pull/18277

**Pull request closed without merging**